File: python/examples/sample_config.py
"""
Sample: Read and write sensor configuration via Config API

Connects to a sensor, reads all config categories, and demonstrates
non-destructive write-back tests.

Usage:
    python sample_config.py [sensor_ip] [config_port]

Copyright 2026 bitsensing Inc. All rights reserved.
"""

import sys
from afi_sdk import AfiSdk, AfiSdkConfig, RangeMode


def main():
    host = sys.argv[1] if len(sys.argv) > 1 else "192.168.10.150"
    port = int(sys.argv[2]) if len(sys.argv) > 2 else 30500

    config = AfiSdkConfig(sensor_ip=host, config_port=port)
    sdk = AfiSdk()

    print(f"Connecting to {host}:{port} ...")
    rc = sdk.init(config)
    if rc != 0:
        print(f"Init failed (rc={rc})")
        return

    # ── Device Info ──
    print("\n=== Device Info ===")

    info = sdk.get_device_info()
    if info:
        print(f"  Serial:     {info.get('serial_number', '?')}")
        print(f"  Model:      {info.get('model_name', '?')}")
        print(f"  HW Version: {info.get('hw_version', '?')}")
        print(f"  SW Version: {info.get('software_version', '?')}")
        print(f"  Hashes:     bl={info.get('bootloader_hash', '?')} "
              f"drv={info.get('driver_hash', '?')} fw={info.get('fw_hash', '?')}")
        print(f"  MAC:        {info.get('mac_address', '?')}")

    status = sdk.get_sensor_status()
    if status:
        print(f"  Op State:   {status.get('operation_state', '?')} "
              f"({status.get('operation_state_code', '?')})")
        print(f"  MCU Temp:   {status.get('temperature_mcu', '?')} C")
        print(f"  RF Temp:    {status.get('temperature_rf', '?')} C")
        print(f"  Voltage:    {status.get('voltage', '?')} V")
        print(f"  Uptime:     {status.get('uptime_sec', '?')} sec")

    diag = sdk.get_diagnostic_info()
    if diag:
        print(f"  Combined:   {diag.get('combined_state', '?')} "
              f"({diag.get('combined_state_code', '?')}), "
              f"active faults: {diag.get('active_fault_count', '?')}")
        for f in diag.get("faults", []):
            print(f"    [{f.get('source', '?')}] {f.get('id', '?')} "
                  f"{f.get('name', '?')} level={f.get('level', '?')} "
                  f"active={f.get('active', '?')}")

    # ── Network ──
    print("\n=== Network Config ===")

    net = sdk.get_network_config()
    if net:
        print(f"  Radar IP:   {net.get('ip_address', '?')}")
        print(f"  Subnet:     {net.get('subnet_mask', '?')}")
        print(f"  Gateway:    {net.get('gateway', '?')}")
        print(f"  API Port:   {net.get('api_port', '?')}")
        for stream in ("detection", "health", "performance"):
            print(f"  {stream.capitalize():11s} {net.get(f'{stream}_ip', '?')}:"
                  f"{net.get(f'{stream}_port', '?')} "
                  f"({net.get(f'{stream}_protocol', '?')})")

    # ── Range Mode ──
    print("\n=== Range Mode ===")

    mode = sdk.get_range_mode()
    if mode is not None:
        print(f"  Mode: {RangeMode.name(mode)} ({mode})")

    # ── Detection ──
    print("\n=== Detection ===")

    thresh = sdk.get_detection_threshold()
    if thresh:
        print(f"  SNR:  {thresh.get('snr_threshold_db', '?')} dB")
        print(f"  CFAR first:  {thresh.get('cfar_table_first_frame', '?')}")
        print(f"  CFAR second: {thresh.get('cfar_table_second_frame', '?')}")

    filters = sdk.get_detection_filters()
    if filters:
        for key, value in sorted(filters.items()):
            if key.startswith("filter_"):
                print(f"  {key[7:]:22s} {'ON' if value else 'OFF'}")

    density = sdk.get_detection_density()
    if density:
        print(f"  range_density: {'ON' if density.get('range_density') else 'OFF'}")
        print(f"  angle_density: {'ON' if density.get('angle_density') else 'OFF'}")

    # ── Time Sync ──
    print("\n=== Time Sync (PTP) ===")

    ptp = sdk.get_ptp_config()
    if ptp:
        print(f"  Profile: {ptp.get('profile', '?')} ({ptp.get('profile_name', '?')})")
        print(f"  State:   {ptp.get('state', '?')}, offset={ptp.get('offset_ns', '?')} ns, "
              f"master={ptp.get('master_clock_id', '?')}")

    # ── Mounting ──
    print("\n=== Mounting ===")

    pos = sdk.get_mounting_position()
    if pos:
        print(f"  Position: ({pos.get('offset_x_m', 0):.3f}, "
              f"{pos.get('offset_y_m', 0):.3f}, "
              f"{pos.get('offset_z_m', 0):.3f}) m")

    rot = sdk.get_rotation()
    if rot:
        print(f"  Rotation: roll={rot.get('roll_deg', 0):.1f} "
              f"pitch={rot.get('pitch_deg', 0):.1f} "
              f"yaw={rot.get('yaw_deg', 0):.1f} deg")

    # ── Persistence ──
    print("\n=== Config Persistence ===")

    cfg_status = sdk.get_config_status()
    if cfg_status:
        print(f"  Load status: {cfg_status.get('load_status_name', '?')} "
              f"({cfg_status.get('load_status', '?')})")

    # ── Write-back Test (non-destructive) ──
    print("\n--- Write-back Test ---")

    if mode is not None:
        rc = sdk.set_range_mode(mode)
        print(f"  SetRangeMode (same value): {'OK' if rc == 0 else f'FAIL ({rc})'}")

    if filters and "filter_fov" in filters:
        rc = sdk.set_detection_filters(filter_fov=filters["filter_fov"])
        print(f"  SetDetectionFilters (same value): {'OK' if rc == 0 else f'FAIL ({rc})'}")

    sdk.close()
    print("\nDone.")


if __name__ == "__main__":
    main()

File: examples/cpp/sample_config.cpp
/**
 * @file sample_config.cpp
 * @brief Sample: Read and write sensor configuration via Config API
 *
 * Usage: sample_config [sensor_ip]
 */
#include <afi/afi_sdk.h>
#include <cstddef>
#include <cstdio>

static void print_result(const char* name, int rc) {
    if (rc == afi::kOk) return;
    printf("  %s: %s (%d)\n", name, afi::AfiErrorStr(rc), rc);
}

static void print_cfar_table(const char* name, const std::vector<uint8_t>& table) {
    printf("  %s: [", name);
    for (std::size_t i = 0; i < table.size(); ++i) {
        printf("%s%u", i == 0 ? "" : ", ", static_cast<unsigned>(table[i]));
    }
    printf("]\n");
}

int main(int argc, char* argv[]) {
    afi::AfiSdkConfig config;
    if (argc > 1) config.sensor_ip = argv[1];
    config.enable_rdi = false;  // Config API only
    config.enable_config_api = true;

    printf("Connecting to %s:%d ...\n", config.sensor_ip.c_str(), config.config_port);

    afi::AfiSdk sdk;
    int rc = sdk.Init(config);
    if (rc != afi::kOk) {
        printf("Init failed: %s (%d)\n", afi::AfiErrorStr(rc), rc);
        return 1;
    }

    // --- Device Info ---
    printf("\n=== Device Info ===\n");

    afi::AfiDeviceInfo dev;
    rc = sdk.GetDeviceInfo(dev);
    print_result("GetDeviceInfo", rc);
    if (rc == afi::kOk) {
        printf("  Serial:     %s\n", dev.serial_number.c_str());
        printf("  Model:      %s\n", dev.model_name.c_str());
        printf("  HW Version: %s\n", dev.hw_version.c_str());
        printf("  SW Version: %s\n", dev.software_version.c_str());
        printf("  Hashes:     bl=%s drv=%s fw=%s\n",
               dev.bootloader_hash.c_str(), dev.driver_hash.c_str(), dev.fw_hash.c_str());
        printf("  MAC:        %s\n", dev.mac_address.c_str());
    }

    afi::AfiSensorStatus status;
    rc = sdk.GetSensorStatus(status);
    print_result("GetSensorStatus", rc);
    if (rc == afi::kOk) {
        printf("  Op State:   %s (%d)\n", status.operation_state.c_str(), status.operation_state_code);
        printf("  MCU Temp:   %.1f C\n", status.temperature_mcu);
        printf("  RF Temp:    %.1f C\n", status.temperature_rf);
        printf("  Voltage:    %.2f V\n", status.voltage);
        printf("  Uptime:     %u sec\n", status.uptime_sec);
    }

    afi::AfiDiagnosticInfo diag;
    rc = sdk.GetDiagnosticInfo(diag);
    print_result("GetDiagnosticInfo", rc);
    if (rc == afi::kOk) {
        printf("  Combined:   %s (%d), active faults: %u\n",
               diag.combined_state.c_str(), diag.combined_state_code,
               diag.active_fault_count);
        for (const auto& f : diag.faults) {
            printf("    [%s] %s %s level=%s active=%d\n",
                   f.source.c_str(), f.id.c_str(), f.name.c_str(),
                   f.level.c_str(), f.active ? 1 : 0);
        }
    }

    // --- Network ---
    printf("\n=== Network Config ===\n");

    afi::AfiNetworkConfig net;
    rc = sdk.GetNetworkConfig(net);
    print_result("GetNetworkConfig", rc);
    if (rc == afi::kOk) {
        printf("  Radar IP:   %s\n", net.ip_address.c_str());
        printf("  Subnet:     %s\n", net.subnet_mask.c_str());
        printf("  Gateway:    %s\n", net.gateway.c_str());
        printf("  API Port:   %d\n", net.api_port);
        printf("  Detection:  %s:%d (%s)\n", net.detection.ip.c_str(),
               net.detection.port, net.detection.protocol.c_str());
        printf("  Health:     %s:%d (%s)\n", net.health.ip.c_str(),
               net.health.port, net.health.protocol.c_str());
        printf("  Perf:       %s:%d (%s)\n", net.performance.ip.c_str(),
               net.performance.port, net.performance.protocol.c_str());
    }

    // --- Range Mode ---
    printf("\n=== Range Mode ===\n");

    afi::AfiRangeMode mode;
    rc = sdk.GetRangeMode(mode);
    print_result("GetRangeMode", rc);
    if (rc == afi::kOk) {
        const char* mode_names[] = {"DR", "LR", "MR", "SR", "ULR", "USR"};
        uint8_t m = static_cast<uint8_t>(mode);
        printf("  Mode: %s (%d)\n", (m < 6) ? mode_names[m] : "Unknown", m);
    }

    // --- Detection ---
    printf("\n=== Detection ===\n");

    afi::AfiDetectionThreshold thresh;
    rc = sdk.GetDetectionThreshold(thresh);
    print_result("GetDetectionThreshold", rc);
    if (rc == afi::kOk) {
        printf("  SNR:  %.1f dB\n", thresh.snr_dB);
        print_cfar_table("CFAR first", thresh.cfar_table_first_frame);
        print_cfar_table("CFAR second", thresh.cfar_table_second_frame);
    }

    afi::AfiDetectionFilters filters;
    rc = sdk.GetDetectionFilters(filters);
    print_result("GetDetectionFilters", rc);
    if (rc == afi::kOk) {
        printf("  FOV:               %s\n", filters.filter_fov ? "ON" : "OFF");
        printf("  APS noise:         %s\n", filters.filter_aps_noise ? "ON" : "OFF");
        printf("  Ego specular:      %s\n", filters.filter_ego_specular ? "ON" : "OFF");
        printf("  Ground sidelobe:   %s\n", filters.filter_ground_sidelobe ? "ON" : "OFF");
        printf("  Upper structure:   %s\n", filters.filter_upper_structure ? "ON" : "OFF");
        printf("  Multibounce 2x:    %s\n", filters.filter_multibounce_2x ? "ON" : "OFF");
        printf("  Struct multipath:  %s\n", filters.filter_structure_multipath ? "ON" : "OFF");
        printf("  Specular mirror:   %s\n", filters.filter_specular_mirror ? "ON" : "OFF");
        printf("  Wheel microdopp.:  %s\n", filters.filter_wheel_microdoppler ? "ON" : "OFF");
    }

    afi::AfiDetectionDensity density;
    rc = sdk.GetDetectionDensity(density);
    print_result("GetDetectionDensity", rc);
    if (rc == afi::kOk) {
        printf("  Range density: %s\n", density.range_density ? "ON" : "OFF");
        printf("  Angle density: %s\n", density.angle_density ? "ON" : "OFF");
    }

    // --- Time Sync ---
    printf("\n=== Time Sync (PTP) ===\n");

    afi::AfiPtpConfig ptp;
    rc = sdk.GetPtpConfig(ptp);
    print_result("GetPtpConfig", rc);
    if (rc == afi::kOk) {
        printf("  Profile: %d (%s)\n", ptp.profile, ptp.profile_name.c_str());
        printf("  State:   %s, offset=%lld ns, master=%s\n",
               ptp.state.c_str(), static_cast<long long>(ptp.offset_ns),
               ptp.master_clock_id.c_str());
    }

    // --- Mounting ---
    printf("\n=== Mounting ===\n");

    afi::AfiMountingPosition pos;
    rc = sdk.GetMountingPosition(pos);
    print_result("GetMountingPosition", rc);
    if (rc == afi::kOk) {
        printf("  Position: (%.3f, %.3f, %.3f) m\n", pos.x, pos.y, pos.z);
    }

    afi::AfiRotation rot;
    rc = sdk.GetRotation(rot);
    print_result("GetRotation", rc);
    if (rc == afi::kOk) {
        printf("  Rotation: roll=%.1f pitch=%.1f yaw=%.1f deg\n",
               rot.roll, rot.pitch, rot.yaw);
    }

    sdk.Stop();
    printf("\nDone.\n");
    return 0;
}
